File: scripts/reorder_models.mjs
import fs from 'fs';
import path from 'path';

const filePath = path.resolve('backend', 'supabase', 'migrations', '20251114000001_seed_data_complete.sql');
const sql = fs.readFileSync(filePath, 'utf8');

const modelPattern = /\(\(SELECT id FROM public\.brands WHERE name = '([^']+)'\), '([^']+)', \(SELECT id FROM public\.vehicle_types WHERE name = '([^']+)'(?: LIMIT 1)?\), ([^,\)]+), ([^\)]+)\)/g;

const entries = [];
const seen = new Set();
let match;

while ((match = modelPattern.exec(sql)) !== null) {
  const [, brand, model, type, yearFrom, yearTo] = match;
  const key = `${brand}||${model}`;
  if (seen.has(key)) continue;
  seen.add(key);
  entries.push({
    brand,
    model,
    type,
    yearFrom: yearFrom.trim(),
    yearTo: yearTo.trim(),
  });
}

if (!entries.length) {
  throw new Error('No se encontraron modelos para reorganizar.');
}

const typeOrder = ['Auto', 'SUV', 'Pickup', 'Van', 'Moto', 'Camión', 'Bus', 'Maquinaria', 'Otro'];
const pluralLabels = {
  Auto: 'Autos',
  SUV: 'SUVs',
  Pickup: 'Pickups',
  Van: 'Vans',
  Moto: 'Motos',
  'Camión': 'Camiones',
  Bus: 'Buses',
  Maquinaria: 'Maquinaria',
  Otro: 'Otros',
};

const grouped = new Map();
typeOrder.forEach((type) => grouped.set(type, []));
entries.forEach((entry) => {
  if (!grouped.has(entry.type)) {
    grouped.set(entry.type, []);
  }
  grouped.get(entry.type).push(entry);
});

for (const [, models] of grouped) {
  models.sort((a, b) => {
    const brandCompare = a.brand.localeCompare(b.brand);
    if (brandCompare !== 0) return brandCompare;
    return a.model.localeCompare(b.model);
  });
}

const escapeSql = (value) => value.replace(/'/g, "''");

const lines = [];
lines.push('-- ===========================================');
lines.push('-- MODELOS POR TIPO DE VEHÍCULO');
lines.push('-- ===========================================');
lines.push('');
lines.push('DO $$');
lines.push('BEGIN');

for (const type of typeOrder) {
  const models = grouped.get(type) || [];
  if (!models.length) continue;
  const label = pluralLabels[type] || `${type}s`;
  lines.push(`    -- ${label}`);
  lines.push('    INSERT INTO public.models (brand_id, name, vehicle_type_id, year_from, year_to) VALUES');
  models.forEach((entry, idx) => {
    const suffix = idx === models.length - 1 ? '' : ',';
    lines.push(
      `        ((SELECT id FROM public.brands WHERE name = '${escapeSql(entry.brand)}'), '${escapeSql(entry.model)}', (SELECT id FROM public.vehicle_types WHERE name = '${entry.type}' LIMIT 1), ${entry.yearFrom}, ${entry.yearTo})${suffix}`
    );
  });
  lines.push('    ON CONFLICT (brand_id, name) DO NOTHING;');
  lines.push('');
}

lines.push('END;');
lines.push('$$ LANGUAGE plpgsql;');
lines.push('');

const block = lines.join('\r\n');
const startMarker = '-- ===========================================\r\n-- MODELOS POR TIPO DE VEHÍCULO';
const endMarker = '-- Asegurar que la columna features';
const startIdx = sql.indexOf(startMarker);
if (startIdx === -1) {
  throw new Error('No se encontró el inicio del bloque de modelos.');
}
const endIdx = sql.indexOf(endMarker, startIdx);
if (endIdx === -1) {
  throw new Error('No se encontró el final del bloque de modelos.');
}
const newSql = sql.slice(0, startIdx) + block + '\r\n' + sql.slice(endIdx);
fs.writeFileSync(filePath, newSql, 'utf8');
console.log(`Se reorganizaron ${entries.length} modelos por tipo de vehículo.`);

File: scripts/reorder_models.py
import re
from pathlib import Path

FILE_PATH = Path(r"c:\Users\chris\OneDrive\Desktop\Simple\backend\supabase\migrations\20251114000001_seed_data_complete.sql")

def load_models(sql: str):
    pattern = re.compile(
        r"\(\(SELECT id FROM public\.brands WHERE name = '([^']+)'\), '([^']+)', \(SELECT id FROM public\.vehicle_types WHERE name = '([^']+)'(?: LIMIT 1)?\), ([^,\)]+), ([^\)]+)\)",
        re.MULTILINE,
    )
    entries = []
    seen = set()
    for match in pattern.finditer(sql):
        brand, model, vtype, year_from, year_to = match.groups()
        key = (brand, model)
        if key in seen:
            continue
        seen.add(key)
        entries.append(
            {
                "brand": brand,
                "model": model,
                "type": vtype,
                "year_from": year_from.strip(),
                "year_to": year_to.strip(),
            }
        )
    return entries


def escape(value: str) -> str:
    return value.replace("'", "''")


def build_insert_block(entries):
    type_order = [
        'Auto',
        'SUV',
        'Pickup',
        'Van',
        'Moto',
        'Camión',
        'Bus',
        'Maquinaria',
        'Otro',
    ]
    grouped = {t: [] for t in type_order}
    for entry in entries:
        grouped.setdefault(entry['type'], []).append(entry)
    for models in grouped.values():
        models.sort(key=lambda x: (x['brand'], x['model']))
    lines = []
    lines.append('-- ===========================================')
    lines.append('-- MODELOS POR TIPO DE VEHÍCULO')
    lines.append('-- ===========================================')
    lines.append('')
    lines.append('DO $$')
    lines.append('BEGIN')
    for vtype in type_order:
        models = grouped.get(vtype) or []
        if not models:
            continue
        lines.append(f"    -- {vtype}s" if vtype not in ['Camión', 'Bus'] else f"    -- {vtype}s")
        lines.append('    INSERT INTO public.models (brand_id, name, vehicle_type_id, year_from, year_to) VALUES')
        row_lines = []
        for entry in models:
            row = (
                "        ((SELECT id FROM public.brands WHERE name = '{brand}'), '{model}', "
                "(SELECT id FROM public.vehicle_types WHERE name = '{vtype}' LIMIT 1), {year_from}, {year_to})"
            ).format(
                brand=escape(entry['brand']),
                model=escape(entry['model']),
                vtype=entry['type'],
                year_from=entry['year_from'],
                year_to=entry['year_to'],
            )
            row_lines.append(row)
        lines.append(',\n'.join(row_lines))
        lines.append('    ON CONFLICT (brand_id, name) DO NOTHING;')
        lines.append('')
    lines.append('END;')
    lines.append('$$ LANGUAGE plpgsql;')
    lines.append('')
    return '\r\n'.join(lines)


def main():
    sql = FILE_PATH.read_text(encoding='utf-8')
    entries = load_models(sql)
    if not entries:
        raise SystemExit('No se encontraron modelos para reorganizar.')
    block = build_insert_block(entries)
    start_marker = '-- ===========================================\r\n-- EJECUCIÓN DEL SEED DATA'
    end_marker = '-- Asegurar que la columna features'
    start_idx = sql.find(start_marker)
    if start_idx == -1:
        raise SystemExit('No se encontró el inicio del bloque de modelos.')
    end_idx = sql.find(end_marker, start_idx)
    if end_idx == -1:
        raise SystemExit('No se encontró el final del bloque de modelos.')
    new_sql = sql[:start_idx] + block + '\r\n' + sql[end_idx:]
    FILE_PATH.write_text(new_sql, encoding='utf-8')
    print(f'Se reorganizaron {len(entries)} modelos por tipo de vehículo.')


if __name__ == '__main__':
    main()
